src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs

```diff
@@ -652,5 +686,84 @@ public Entry(ProjectState state)
                 State = state;
             }
         }
+
+        private class ProjectContainer : Dictionary<ProjectKey, Entry>
+        {
+            public bool ContainsKey(ProjectId projectId)
+            {
+                foreach (var item in this)
+                {
+                    if (item.Key.Id == projectId)
+                    {
+                        return true;
+                    }
+                }
+
+                return false;
+            }
+
+            public bool TryGetValue(string filePath, out Entry entry)
+            {
+                foreach (var item in this)
+                {
+                    if (FilePathComparer.Instance.Equals(item.Key.FilePath, filePath))
+                    {
+                        entry = item.Value;
+                        return true;
+                    }
+                }
+
+                entry = null;
+                return false;
+            }
+
+            public bool TryGetValue(ProjectId projectId, out Entry entry)
+            {
+                foreach (var item in this)
+                {
+                    if (item.Key.Id == projectId)
+                    {
+                        entry = item.Value;
+                        return true;
+                    }
+                }
+
+                entry = null;
+                return false;
+            }
+        }
+
+        internal struct ProjectKey : IEquatable<ProjectKey>
```

Comment:
I have some concerns about this, because I expected that the projectId was **fundamental**. When would you ever have two projects with the same ID but different file paths?

I want to resolve this issue before getting too much further because I think this is central.

Reply:
Glad you brought that up. That's a large part of why I sent out a design PR first, didn't feel super confident in the entire direction.

So, it's not that you'd ever have two projects with same id but different file paths but more that you need to be able to look up a project via project id or file path because there are scenarios when you only have one of those two. Aka, I technically could have had two separate dictionaries tracking the associations with projects and that'd be the same. I just tried to normalize it here so we didn't have to do double the operations in cases where entries changed.

Reply:
What would happen if you just didn't do any of that? linear search?

Looking up a project by filename isn't **fundamental** - it's not something we should be doing frequently.

Reply:
Ya I could, just felt a little rough iterating through and then checking `item.Value.State.HostProject.FilePath`. Didn't feel strongly.

Reply:
SUre

Reply:
> What is the case where we have a filepath but not an Id?

Roslyn Project => Razor Project lookup